src/Sarif/Visitors/RunMergingVisitor.cs

```diff
@@ -78,7 +78,23 @@ public override Result VisitResult(Result node)
             }
 
             // Cache RuleId and set Result.RuleIndex to the (new) index
-            string ruleId = node.ResolvedRuleId(CurrentRun);
+            string resultRuleId = node.ResolvedRuleId(CurrentRun);
+            string ruleId = resultRuleId;
+            ReportingDescriptor rule = null;
+
+            // Result's rule Id is a sub rule Id
+            if (!string.IsNullOrEmpty(ruleId) &&
+                ruleId.IndexOf(SarifConstants.HierarchicalComponentSeparator) > 0)
+            {
+                rule = node.GetRule(CurrentRun);
+                string possilbeParentRuleId = rule?.Id;
+                if (possilbeParentRuleId != null &&
+                    ruleId.StartsWith(possilbeParentRuleId + SarifConstants.HierarchicalComponentSeparator))
+                {
+                    ruleId = possilbeParentRuleId;
+                }
+            }
+
             if (!string.IsNullOrEmpty(ruleId))
             {
                 if (RuleIdToIndex.TryGetValue(ruleId, out int ruleIndex))
```

Comment:
>RuleIdToIndex [](http://example.com/codeflow?start=20&length=13)

It uses this rule-id to index mapping to track all referenced rules. 
If there are multiple unique sub ruleIds of a same parent rule in the results, it caches the same parent rule for each unique sub ruleId, which caused the invalid Sarif output.